drv/ignition-server/src/main.rs

```diff
@@ -0,0 +1,243 @@
+        port: u8,
+    ) -> Result<PortState, RequestError> {
+        if port >= self.port_count {
+            return Err(RequestError::from(IgnitionError::InvalidPort));
+        }
+
+        self.port_state(port).map_err(RequestError::from)
+    }
+
+    fn counters(
+        &mut self,
+        _: &userlib::RecvMessage,
+        port: u8,
+    ) -> Result<Counters, RequestError> {
+        if port >= self.port_count {
+            return Err(RequestError::from(IgnitionError::InvalidPort));
+        }
+
+        self.controller
+            .counters(port)
+            .map_err(IgnitionError::from)
+            .map_err(RequestError::from)
+    }
+
+    fn link_events(
+        &mut self,
+        _: &userlib::RecvMessage,
+        port: u8,
+        link: LinkSelect,
+    ) -> Result<LinkEvents, RequestError> {
+        if port >= self.port_count {
+            return Err(RequestError::from(IgnitionError::InvalidPort));
+        }
+
+        self.controller
+            .link_events(port, link)
+            .map_err(IgnitionError::from)
+            .map_err(RequestError::from)
+    }
+
+    fn clear_link_events(
+        &mut self,
+        _: &userlib::RecvMessage,
+        port: u8,
+        link: LinkSelect,
+    ) -> Result<(), RequestError> {
+        if port >= self.port_count {
+            return Err(RequestError::from(IgnitionError::InvalidPort));
+        }
+
+        self.controller
+            .clear_link_events(port, link)
+            .map_err(IgnitionError::from)
+            .map_err(RequestError::from)
+    }
+
+    fn send_request(
+        &mut self,
+        _: &userlib::RecvMessage,
+        port: u8,
+        request: Request,
+    ) -> Result<(), RequestError> {
+        if port >= self.port_count {
+            return Err(RequestError::from(IgnitionError::InvalidPort));
+        }
+
+        let port_state = self.port_state(port).map_err(RequestError::from)?;
+        let target_state = port_state
+            .target()
+            .ok_or(RequestError::from(IgnitionError::NoTargetPresent))?;
+
+        if target_state.system_reset_in_progress()
+            || target_state.system_power_off_in_progress()
+            || target_state.system_power_on_in_progress()
+        {
+            return Err(RequestError::from(IgnitionError::RequestInProgress));
+        }
+
+        self.controller
+            .set_request(port, request)
+            .map_err(IgnitionError::from)
+            .map_err(RequestError::from)
+    }
+}
+
+impl idol_runtime::NotificationHandler for ServerImpl {
+    fn current_notification_mask(&self) -> u32 {
+        TIMER_NOTIFICATION_MASK
+    }
+
+    fn handle_notification(&mut self, _bits: u32) {
+        let start = sys_get_timer().now;
+
+        // Only poll the presence summary if the port count seems reasonable. A
+        // count of 0xff may occur if the FPGA is running an incorrect
+        // bitstream.
+        if self.port_count > 0 && self.port_count != 0xff {
```

Comment:
Should the `port_count()` idol function return an error if the port count is `0xff` too?